File: Week3_Intro_HashTables/majority-element.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        # Initialize frequency table
        freq = {}
        # Threshold that must be exceeded based on question
        majority_threshold = len(nums) // 2
        
        for i in nums:
            # This code pattern is very, very common!
            if i in freq:
                freq[i] += 1
            else:
                freq[i] = 1
            
            # Check if this is indeed majority element
            if freq[i] > majority_threshold:
                return i
        
        # follow up for 2-pass method:
        # for k, v in freq.items():
            # if v > majority_threshold:
                # return k

File: Session1_Intro_HashTables/majority-element.java
import java.util.*;

class Solution {
    public int majorityElement(int[] nums) {
        /* Initialize HashMap */
        Map<Integer, Integer> freq = new HashMap<>();

        /* Majority Threshold defined in the question */
        int majorityThreshold = nums.length / 2;

        for (int i : nums) {
            /* Java's pattern for finding count */
            freq.put(i, freq.getOrDefault(i, 0) + 1);
            /* Check whether majority element */
            if (freq.get(i) > majorityThreshold)
                return i;
        }

        /* Dead line of code since majority element is guaranteed */
        return -1;
    }
}